Sources/JS/initTypeCollection.js

```diff
@@ -12,89 +12,71 @@
-    if(this.group != null) result.group = this.group
-    if(this.properties != null) result.properties = this.properties.slice()
-    if(this.methods != null) result.properties = this.methods.slice()
-    if(this.signature != null) result.signature = this.signature.serialize()
+Type.prototype.union = function(otherType) {
+    var newType = new Type()
+    newType.definiteType = this.definiteType & otherType.definiteType
+    newType.possibleType = this.possibleType | otherType.possibleType
 
-    return result
-}
-
-function FunctionSignature() {}
-FunctionSignature.prototype.serialize = function() {
-    var result = {}
-    if(this.inputTypes != null) {
-        result.inputTypes = []
-        for (const inputType in this.inputTypes) {
-            result.inputTypes.append(inputType.serialize())
-        }
-    }
-    if(this.outputType != null)  result.outputType = this.outputType.serialize()
+    if (this.group === otherType.group) newType.group = this.group
+    newType.properties = arrayIntersection(this.properties, otherType.properties)
+    newType.methods = arrayIntersection(this.methods, otherType.methods)
 
-    return result
+    return newType
 }
 
 // variableNumber: Type
 var types = {}
-// Must be kept in sync with TypeSystem.swift
-baseTypes = {
-    nothing: 0,
-    undefined: 1 << 0,
-    integer: 1 << 1,
-    float: 1 << 2,
-    string: 1 << 3,
-    boolean: 1 << 4,
-    object: 1 << 5,
-    function: 1 << 6,
-    constructor: 1 << 7,
-    unknown: 1 << 8,
-    bigint: 1 << 9,
-    regexp: 1 << 10,
-}
-function updateType(number, value){
-    function setType(rawValue) {
-        types[number] = new Type()
-        types[number].definiteType = rawValue
-        types[number].possibleType = rawValue
+function getBaseType(rawValue) {
+    var newType = new Type()
+    newType.definiteType = rawValue
+    newType.possibleType = rawValue
+    if (rawValue === baseTypes.object) {
+        newType.properties = []
+        newType.methods = []
     }
+    return newType
+}
+function getCurrentType(value){
     try {
         // Fuzzilli handles both null/undefined as undefined
-        if (value == null) {
-            setType(baseTypes.undefined)
-            return
-        }
-        if (Number.isInteger(value)) {
-            setType(baseTypes.integer)
-            return
-        }
-        if (typeof value === 'number') {
-            setType(baseTypes.float)
-            return
-        }
-        if (typeof value === 'string') {
-            setType(baseTypes.string)
-            return
-        }
-        if (typeof value === 'boolean') {
-            setType(baseTypes.boolean)
-            return
-        }
-        if (typeof value === 'bigint') {
-            setType(baseTypes.bigint)
-            return
-        }
+        if (value == null) return getBaseType(baseTypes.undefined)
+        if (Number.isInteger(value)) return getBaseType(baseTypes.integer)
```

Comment:
Maybe this needs to be wrapped in try-catch as well since script code could redefine `Number.isInteger`?

Reply:
Then it would match all integers as floats, are we okay with that?

Reply:
I think a fairly robust approach would be to
1. "Backup" the `Number.isInteger` function (as well as other required runtime functions) in `initTypeCollectionScript` e.g. by doing something like `var NumberIsInteger = Number.isInteger;`
2. In here, use `NumberIsInteger` in a try catch
3. If it fails, either use unknown or, in this particular case, return .float, which is also fine

Then it should only fail in the rare case that script somehow reassigns the global `NumberIsInteger` variable. 
If you wanted to make it even more robust, you could install all type collection helper variables via `Object.defineProperty(this, '$VAR_NAME', { value: $VALUE, writable: false, configurable: false, enumerable: false}` then they should basically be invisible and immutable but still be available simply via their name:

```
Object.defineProperty(this, 'foobar', {value: 42, writable: false, configurable: false, enumerable: false})
foobar
// 42
foobar = 43
foobar
// 42
delete this.foobar
foobar
// 42
```

You could also group all these global variables together under a `typeCollectionContext` object.

Reply:
I guess the first one option is enough, so we don'ot overcomplicate it